sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1644 @@
+
+**Rationale:** This method makes it possible to set values of a sequence type
+without incurring the costs of splitting the sequence into subsequences,
+and concatenating them back together with the new value.  This operation is also
+a common facility in many other languages, so supporting it in Clarity would
+make Clarity more approachable to new developers.
+
+**Examples:**
+
+```clarity
+(replace-at u"ab" u1 u"c") ;; Returns (some u"ac")
+(replace-at 0x00112233 u2 0x44) ;; Returns (some 0x00114433)
+(replace-at "abcd" u3 "e") ;; Returns (some "abce")
+(replace-at (list 1) u0 10) ;; Returns (some (10))
+(replace-at (list (list 1) (list 2)) u0 (list 33)) ;; Returns (some ((33) (2)))
+(replace-at (list 1 2) u3 4) ;; Returns none
+```
+
+### New global: `tx-sponsor?`
+
+* **Type:** `(optional principal)`
+
+This global variable evaluates to the fee-sponsoring principal of the current transaction (if there is such a principal).
+
+**Rationale:** Stacks already exposes the origin address as `tx-sender`, and
+exposing this data would make it possible to build smart contracts where only
+certain entities could relay transactions on behalf of users (such as entities
+with which the user has an off-chain business relationship).
+
+### New global: `chain-id`
+
+* **Type:** `uint`
+
+This global variable evaluates to a 32-bit chain ID that identifies this
+instance of the Stacks blockchain.  The purpose of this global variable is to
+give contract developers a way to determine which instance of Stacks their code
+runs on, in order to execute instance-specific logic or integrate with
+instance-specific off-chain services.
+
+The following values are reported:
+
+* `u1`:  This is the Stacks mainnet chain (this is `0x00000001`).
+
+* `u2147483648`:  This is the Stacks testnet chain (this is `0x80000000`).
+
+Other values may be used for other deployments of Stacks, such as (but not
+limited to) subnets.
+
+**Rationale**: Stacks already has two instances (testnet and mainnet), and
+subnets are instances of Stacks.  Enabling smart contracts to differentiate
+between which environments they run in would empower developers to add
+(or rely on) chain-specific features.
+
+### New feature: Faster Clarity Parser
+
+This SIP proposes a new Clarity lexer and parser implementation that will be
+specific to Clarity 2.  Current benchmarks indicate that the reference
+implementation of this new lexer and parser it is around 3x faster than the
+reference implementation of the chain today.
+
+The reason for proposing this as a breaking change is to de-risk the possibility
+that the new lexer and parser are not "bug-for-bug" compatible with the current
+parser.  Tying the release of this new lexer and parser with a breaking change,
+while keeping the old parser for processing all blocks prior to the change,
+ensures that any new behaviors introduced will only apply to transactions after
+the change takes effect.
```

Comment:
Can you list all new behaviors this change introduce as well as all old behaviors that has been dropped with rationale behind these decisions?

Reply:
Given the current testing regiment, I tentatively believe (but cannot prove) that the new implementation behaves the same as the old implementation.  However, the fact that I cannot prove that they are equivalent makes shipping the new parser in anything other than a hard fork risky -- if it turns out that my belief was wrong, and someone is able to produce a Clarity contract that parses _differently_ with these two versions, then we risk a chain split.  Old nodes with the old parser will not treat the transaction the same away as new nodes with the new parser.

This text is saying that by shipping the new parser as part of a breaking change, we are de-risking this possibility of an accidental chain split.

Reply:
>  if it turns out that my belief was wrong, and someone is able to produce a Clarity contract that parses differently with these two versions, then we risk a chain split.

To support the point @jcnelson is making, there is indeed Clarity code that parses in the original parser but shouldn't parse in the upgrade, such as:

```clarity
(list "\\" x")
```

See https://github.com/stacks-network/stacks-blockchain/issues/3123

Reply:
The [current V1 parser](https://github.com/stacks-network/stacks-blockchain/blob/1a34c115089434e3f23abf198122cd5d58b4c921/clarity/src/vm/ast/parser/mod.rs) is around 700 LOC (ignoring unit tests) while the [lexer and parser for Stacks 2.1](https://github.com/stacks-network/stacks-blockchain/tree/next/clarity/src/vm/ast/parser/v2) is almost 2000 LOC (without unit tests) for V2 _plus_ 700 LOC for the V1 parser.

For all  the extra complexity with the parser upgrade, there better be other advances than that the new parser is "3x faster".

Reply:
Thanks @njordhov for pointing this bug fix.
In interpreted languages parser is the cornerstone of everything. And even a small change can cause a cascade of unknown behaviors/quirks. That is why I wanted to have a list of what has been fixed, added, removed or modified.

Reply:
> For all the extra complexity with the parser upgrade, there better be other advances than that the new parser is "3x faster".

@obycode can speak to this better than I can (he wrote the new parser), but from memory, the new parser:

* Separates the task of lexing the source code into tokens from the task of parsing the tokens into symbolic expressions.  This makes it possible to rigorously test both passes in isolation.

* Continues parsing the program even after errors are encountered, so the full list of errors can be reported to the developer (the v1 parser only reports the first error encountered).

* Parses comments as first-class tokens, instead of ignoring them.  This empowers tools like Clarinet to use comments as a way for embedding extra metadata to Clarity code, such as for auto-generating documentation or performing additional static analysis passes.

Reply:
Yes, @jcnelson's points above are all addition benefits of this change. This new parser is actually 8x faster than the old parser. The biggest contributing factor to the higher lines of code count is also a massive improvement to the readability of the code, that is the change away from regular expressions. The regular expressions in the old parser make the code much more difficult to reason about (in addition to making it much slower).

Reply:
@obycode has anything changed with regards to:
 - handling white spaces (in V1 they were mandatory in like 98% cases)
 - handling text encoding (can we write contracts in any other encoding than utf8? what about line endings?)
 - handling keywords (in V1 we can easily overwrite multiple keywords and change their behavior/value)
 - handling optional traits
 ?

Reply:
Yes, thanks @lnow.
* handling white spaces (in V1 they were mandatory in like 98% cases)
  - There are some small changes to whitespace, but the previous behavior is largely left as-is
  - `{a:1}` previously gave a "whitespace expected..." error, but is now legal
* handling text encoding (can we write contracts in any other encoding than utf8? what about line endings?)
  - It is still the case that only ASCII characters are allowed in the contract
* handling keywords (in V1 we can easily overwrite multiple keywords and change their behavior/value)
  - This is handled in the type-checker, not in the parser, but no changes were made, developers can still overwrite some keywords, same as before)
* handling optional traits
  - No changes in the parser for this, but see the other comments in this PR about improvements to traits

Additionally:
* Fixes an issue where a tuple in a comment shows up in the AST (stacks-network/stacks-blockchain#3124)
* Escaped backslash in strings may fail parsing (stacks-network/stacks-blockchain#3123)